tests/host_tools/memory.py

```diff
@@ -137,12 +126,42 @@ def run(self):
 
             time.sleep(self.MEMORY_SAMPLE_TIMEOUT_S)
 
-    def is_in_guest_mem_region(self, address):
-        """Check if the address is inside the guest memory region."""
-        if self._guest_mem_start is None:
-            return False
-        guest_mem_end = self._guest_mem_start + self.guest_mem_mib
-        return self._guest_mem_start <= address < guest_mem_end
+    def update_guest_mem_regions(self, address, size_kib):
+        """
+        If the address is recognised as a guest memory region,
+        cache it and return True, otherwise return False.
+        """
+
+        # If x86_64 guest memory exceeds 3328M, it will be split
+        # in 2 regions: 3328M and the rest. We have 3 cases here
+        # to recognise a guest memory region:
+        #  - its size matches the guest memory exactly
+        #  - its size is 3328M
+        #  - its size is guest memory minus 3328M.
+        if size_kib in (
+            self.guest_mem_mib * 1024,
+            self.X86_MEMORY_GAP_START,
+            self.guest_mem_mib * 1024 - self.X86_MEMORY_GAP_START,
+        ):
+            if not self._guest_mem_start_1:
+                self._guest_mem_start_1 = address
+                return True
+            if not self._guest_mem_start_2:
+                self._guest_mem_start_2 = address
+                return True
+        return False
+
+    def is_in_guest_mem_regions(self, address):
+        """Check if the address is inside a guest memory region."""
+        if self._guest_mem_start_1:
+            guest_mem_end = self._guest_mem_start_1 + self.guest_mem_mib
+            if self._guest_mem_start_1 <= address < guest_mem_end:
+                return True
+        if self._guest_mem_start_2:
+            guest_mem_end = self._guest_mem_start_2 + self.guest_mem_mib
+            if self._guest_mem_start_2 <= address < guest_mem_end:
+                return True
+        return False
```

Comment:
Optional: consider using ranges here:

```
In [18]: r = range(1000, 10_000)

In [19]: 9_999 in r
Out[19]: True

In [20]: 10_000 in r
Out[20]: False
```

so it would look like:

```
for guest_range in [self._guest_range_1, self._guest_range_2]:
    if guest_range is not None and address in guest_range:
        return True
return False
```

One thing I don't understand is that both ranges are the same size (`self.guest_mem_mib`)? Shouldn't they have different sizes?

Reply:
> Also, if you change "Related: #3349" to "Resolves: #3349", it will automatically resolve the issue once this gets merged.

Thanks, didn't know that.